gcc: fix c++ headers when same triplet cross compiling

uri-canva opened the pull request:
###### Description of changes

When build platform and host platform differ, but have the same triple, the code in nixpkgs will consider it a cross compilation, but gcc won't. This will lead some derivations to look for c++ headers in the wrong place. To solve this always output the headers in the non-cross location, like we do for the other gcc headers already.

Before:
```
$ nix-build -A pkgsCross.gnu64.boehmgc
...
gc_badalc.cc:29:10: fatal error: new: No such file or directory
   29 | #include <new> // for bad_alloc, precedes include of gc_cpp.h
      |          ^~~~~
compilation terminated.
gc_cpp.cc:36:10: fatal error: new: No such file or directory
   36 | #include <new> // for bad_alloc, precedes include of gc_cpp.h
      |          ^~~~~
compilation terminated.
make[1]: *** [Makefile:1780: gc_badalc.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:1780: gc_cpp.lo] Error 1
In file included from extra/gc.c:74:
In function 'GC_suspend_handler_inner',
    inlined from 'GC_suspend_handler' at extra/../pthread_stop_world.c:261:7:
extra/../pthread_stop_world.c:359:19: warning: '__atomic_load_8' writing 8 bytes into a region of size 0 overflows the destination [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wstringop-overflow=-Wstringop-overflow=8;;]
  359 |     suspend_cnt = (word)ao_load_async(&(me -> stop_info.ext_suspend_cnt));
make[1]: Leaving directory '/build/gc-8.2.2'
make: *** [Makefile:1933: all-recursive] Error 1
error: builder for '/nix/store/3628wviya1i588nyp208lf3dhh1kchzr-boehm-gc-x86_64-unknown-linux-gnu-8.2.2.drv' failed with exit code 2;
       last 10 log lines:
       > make[1]: *** [Makefile:1780: gc_badalc.lo] Error 1
       > make[1]: *** Waiting for unfinished jobs....
       > make[1]: *** [Makefile:1780: gc_cpp.lo] Error 1
       > In file included from extra/gc.c:74:
       > In function 'GC_suspend_handler_inner',
       >     inlined from 'GC_suspend_handler' at extra/../pthread_stop_world.c:261:7:
       > extra/../pthread_stop_world.c:359:19: warning: '__atomic_load_8' writing 8 bytes into a region of size 0 overflows the destination [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wstringop-overflow=-Wstringop-overflow=8;;]
       >   359 |     suspend_cnt = (word)ao_load_async(&(me -> stop_info.ext_suspend_cnt));
       > make[1]: Leaving directory '/build/gc-8.2.2'
       > make: *** [Makefile:1933: all-recursive] Error 1
       For full logs, run 'nix-store -l /nix/store/3628wviya1i588nyp208lf3dhh1kchzr-boehm-gc-x86_64-unknown-linux-gnu-8.2.2.drv'.
```

After:
```
$ nix-build -A pkgsCross.gnu64.boehmgc
...
/nix/store/ddfdqk6zi5rag4nvp8702g9frv6v2i1g-boehm-gc-x86_64-unknown-linux-gnu-8.2.2
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Closes #243166

uri-canva committed:
gcc: fix c++ headers when same triplet cross compiling

When build platform and host platform differ, but have the same
triple, the code in nixpkgs will consider it a cross compilation,
but gcc won't. This will lead some derivations to look for c++
headers in the wrong place. To solve this always output the headers
in the non-cross location, like we do for the other gcc headers
already.

Artturin commented on pkgs/development/compilers/gcc/common/configure-flags.nix:
```diff
@@ -135,6 +135,8 @@ let
       # as a native case.
       "--with-build-sysroot=/"
+      # Same with the stdlibc++ headers embedded in the gcc output
+      "--with-gxx-include-dir=${placeholder "out"}/include/c++/12.3.0/"
```
the version will change, `${version}` should be correct

Artturin commented:
What build does this fix?
im still getting
`bash> configure: error: C compiler cannot create executables`
in `pkgsCross.gnu64.bash`

amjoseph-nixpkgs commented:
Could you provide a test case?

I do tons of builds where "build platform and host platform differ, but have the same triple" because I customize my compiler flags by modifying the `hostPlatform`.  I had to fix a few bugs (the fixes are merged) to make that work, but I guess I missed one.

amjoseph-nixpkgs commented:
@ofborg eval

amjoseph-nixpkgs commented:
Alright, first things first, a complete bug report describing the issue that this fixes: #243164

amjoseph-nixpkgs commented:
> Both `nix-build -A pkgsCross.gnu64-with-xcode.bash`

We need test cases that work for people without darwin systems.

> and `nix-build -A pkgsCross.gnu64.bash` fail with:

That failure is due to #40797 which is fixed by #238154.

amjoseph-nixpkgs commented:
I apologize for taking so long to review this.

I'm rebuilding my environment with this PR; if all goes well I will merge it right after 23.11 branches off.

amjoseph-nixpkgs commented:
> Awesome! I saw your recent change to compare platforms instead of triples (#264976), that's related to the same "platforms being different but triples being the same issue" right?

Yes; that, plus this, mostly solves the problem.

There are a bunch of small package-specific fixes too, like these:

- https://github.com/NixOS/nixpkgs/pull/265759
- https://github.com/NixOS/nixpkgs/pull/265162

amjoseph-nixpkgs commented:
Also there is a test case for this now in `tests.cross.sanity`:

- https://github.com/NixOS/nixpkgs/pull/264992

It's commented out now, but once it passes I will uncomment it.  So when this gets fixed it will stay fixed.

Thank you again for your patience.

lilyball commented:
It sounds like this will fix https://github.com/NixOS/nixpkgs/issues/264865, which should probably be considered a dupe of your https://github.com/NixOS/nixpkgs/issues/243166 but is a lot more explicit about what's going on.

amjoseph-nixpkgs commented:
So, I have good news and bad news.

The good news is that everything but one package builds properly with this PR.

The bad news is that package is `firefox`.  It's a huge build, with a ton of dependencies, so just reproducing the failure takes hours.

Any ideas on what's going on here?  The failure is posted below (for `pkgsCross.aarch64-multiplatform.firefox`).  Apparently after this PR `firefox`'s `clang` can't find `cstddef`.  But nothing other than firefox seems to malfunction.

```
firefox-unwrapped>  0:02.01(B checking the host C compiler works... yes(B(B
firefox-unwrapped>  0:02.01(B checking for the host C++ compiler... /nix/store/1asqji9djmdlapzs70q7jw2j308ry7cn-clang-wrapper-16.0.6/bin/c++(B(B
firefox-unwrapped>  0:02.14(B checking whether the host C++ compiler can be used... yes(B(B
firefox-unwrapped>  0:02.14(B checking the host C++ compiler version... 16.0.6(B(B
firefox-unwrapped>  0:02.21(B checking the host C++ compiler works... yes(B(B
firefox-unwrapped>  0:02.40(B checking for target linker... lld(B(B
firefox-unwrapped>  0:02.51(B checking for host linker... lld(B(B
firefox-unwrapped>  0:02.60(B checking for 64-bit OS... yes(B(B
firefox-unwrapped>  0:02.67(B checking for new enough STL headers from libstdc++...(B(B
firefox-unwrapped>  0:02.67(B DEBUG: <truncated - see config.log for full output>(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |                 #if defined(__GLIBCXX__) && !defined(_GLIBCXX_RELEASE)(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #  error libstdc++ not new enough(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #endif(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #if defined(_GLIBCXX_RELEASE)(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #  if _GLIBCXX_RELEASE < 8(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #    error libstdc++ not new enough(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #  else(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |      (void) 0(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #  endif(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #endif(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |                   ;(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |                   return 0;(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |                 }(B(B
firefox-unwrapped>  0:02.67(B DEBUG: Executing: `/nix/store/7v4bi4q334yircaznwm353h1l5i7k98f-aarch64-unknown-linux-gnu-clang-wrapper-16.0.6/bin/aarch64-unknown-linux-gnu-clang++ /build/conftest.qvbo58dv.cpp -c`(B(B
firefox-unwrapped>  0:02.67(B DEBUG: The command returned non-zero exit status 1.(B(B
firefox-unwrapped>  0:02.67(B DEBUG: Its error output was:(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | /build/conftest.qvbo58dv.cpp:1:10: fatal error: 'cstddef' file not found(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #include <cstddef>(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |          ^~~~~~~~~(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | 1 error generated.(B(B
firefox-unwrapped>  0:02.67(B ERROR: The libstdc++ in use is not new enough.  Please run ./mach bootstrap to update your compiler, or update your system libstdc++ installation.(B(B
firefox-unwrapped> *** Fix above errors and then restart with "./mach build"
error: build of '/nix/store/5in7xkji5hzqkl14ygwq3vxnni54lykk-firefox-unwrapped-aarch64-unknown-linux-gnu-119.0.1.drv' on 'ssh://root@192.168.22.103' failed: builder for '/nix/store/5in7xkji5hzqkl14ygwq3vxnni54lykk-firefox-unwrapped-aarch64-unknown-linux-gnu-119.0.1.drv' failed with exit code 1
error: builder for '/nix/store/5in7xkji5hzqkl14ygwq3vxnni54lykk-firefox-unwrapped-aarch64-unknown-linux-gnu-119.0.1.drv' failed with exit code 1;
```

amjoseph-nixpkgs commented:
Also, @uri-canva could you put this in the git comment?

> When build platform and host platform differ, but have the same triple, the code in nixpkgs will consider it a cross compilation, but gcc won't. This will lead some derivations to look for c++ headers in the wrong place. To solve this always output the headers in the non-cross location, like we do for the other gcc headers already.

amjoseph-nixpkgs commented:
> Any ideas on what's going on here?

Ooh, I bet the clang-specific junk in `cc-wrapper` needs to be updated for this.

Commit:
cc-wrapper: drop no-longer-necessary hack

The commit prior to this one, "gcc: fix c++ headers when same
triplet cross compiling" causes gcc's c++ headers to be in the same
outpath subdirectory regardless of whether the gcc build is a
host==target or host!=target compiler.

As a result of that change, the hack in cc-wrapper which adapted to
the different paths is no longer needed.  And, in fact, it must be
removed, since if it is left in place builds such as
pkgsCross.aarch64-multiplatform.firefox will fail as shown below.

```
firefox-unwrapped>  0:02.01(B checking the host C compiler works... yes(B(B
firefox-unwrapped>  0:02.01(B checking for the host C++ compiler... /nix/store/1asqji9djmdlapzs70q7jw2j308ry7cn-clang-wrapper-16.0.6/bin/c++(B(B
firefox-unwrapped>  0:02.14(B checking whether the host C++ compiler can be used... yes(B(B
firefox-unwrapped>  0:02.14(B checking the host C++ compiler version... 16.0.6(B(B
firefox-unwrapped>  0:02.21(B checking the host C++ compiler works... yes(B(B
firefox-unwrapped>  0:02.40(B checking for target linker... lld(B(B
firefox-unwrapped>  0:02.51(B checking for host linker... lld(B(B
firefox-unwrapped>  0:02.60(B checking for 64-bit OS... yes(B(B
firefox-unwrapped>  0:02.67(B checking for new enough STL headers from libstdc++...(B(B
firefox-unwrapped>  0:02.67(B DEBUG: <truncated - see config.log for full output>(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |                 #if defined(__GLIBCXX__) && !defined(_GLIBCXX_RELEASE)(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #  error libstdc++ not new enough(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #endif(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #if defined(_GLIBCXX_RELEASE)(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #  if _GLIBCXX_RELEASE < 8(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #    error libstdc++ not new enough(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #  else(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |      (void) 0(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #  endif(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #endif(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |                   ;(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |                   return 0;(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |                 }(B(B
firefox-unwrapped>  0:02.67(B DEBUG: Executing: `/nix/store/7v4bi4q334yircaznwm353h1l5i7k98f-aarch64-unknown-linux-gnu-clang-wrapper-16.0.6/bin/aarch64-unknown-linux-gnu-clang++ /build/conftest.qvbo58dv.cpp -c`(B(B
firefox-unwrapped>  0:02.67(B DEBUG: The command returned non-zero exit status 1.(B(B
firefox-unwrapped>  0:02.67(B DEBUG: Its error output was:(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | /build/conftest.qvbo58dv.cpp:1:10: fatal error: 'cstddef' file not found(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | #include <cstddef>(B(B
firefox-unwrapped>  0:02.67(B DEBUG: |          ^~~~~~~~~(B(B
firefox-unwrapped>  0:02.67(B DEBUG: | 1 error generated.(B(B
firefox-unwrapped>  0:02.67(B ERROR: The libstdc++ in use is not new enough.  Please run ./mach bootstrap to update your compiler, or update your system libstdc++ installation.(B(B
firefox-unwrapped> *** Fix above errors and then restart with "./mach build"
error: build of '/nix/store/5in7xkji5hzqkl14ygwq3vxnni54lykk-firefox-unwrapped-aarch64-unknown-linux-gnu-119.0.1.drv' on 'ssh://root@192.168.22.103' failed: builder for '/nix/store/5in7xkji5hzqkl14ygwq3vxnni54lykk-firefox-unwrapped-aarch64-unknown-linux-gnu-119.0.1.drv' failed with exit code 1
error: builder for '/nix/store/5in7xkji5hzqkl14ygwq3vxnni54lykk-firefox-unwrapped-aarch64-unknown-linux-gnu-119.0.1.drv' failed with exit code 1;
```

amjoseph-nixpkgs reviewed (approved):
Good to go!

amjoseph-nixpkgs commented:
@ofborg eval

`aws-sdk-cpp` tests are flaky *again*...

amjoseph-nixpkgs commented:
Rebased to force ofborg to re-eval.

amjoseph-nixpkgs commented:
Alright, I will merge this as soon as `staging` has branched off post-23.11.

lilyball commented:
We're in december, can this be merged yet?

amjoseph-nixpkgs commented:
> We're in december, can this be merged yet?

Sorry about the delay!
